marlowe-chain-sync/libchainsync/Language/Marlowe/Runtime/ChainSync/NodeClient.hs

```diff
@@ -0,0 +1,272 @@
+  -> QueryInMode CardanoMode result
+  -> IO (Either AcquiringFailure result)
+
+
+data NodeClient = NodeClient
+  {
+    submitTxToNode  :: SubmitToNode
+  , queryNode :: forall result . QueryNode result
+  }
+
+
+data NodeClientDependencies r = NodeClientDependencies
+  {
+    connectToLocalNode :: !(LocalNodeClientProtocolsInMode CardanoMode -> IO ())
+  , eventBackend :: !(EventBackend IO r NodeClientSelector)
+  }
+
+
+data NodeClientSelector f where
+  Connect :: NodeClientSelector Void
+  Submit :: NodeClientSelector (TxInMode CardanoMode)
+  Query :: NodeClientSelector (Q.Some (QueryInMode CardanoMode))
+
+
+getNodeClientSelectorConfig :: GetSelectorConfig NodeClientSelector
+getNodeClientSelectorConfig Connect = SelectorConfig "connect" True absurdFieldConfig
+getNodeClientSelectorConfig Submit =
+  SelectorConfig "submit" True $ singletonFieldConfigWith (SomeJSON . txToJSON) "tx" True
+  where
+    txToJSON :: TxInMode CardanoMode -> Value
+    txToJSON (TxInMode tx ShelleyEraInCardanoMode) = toJSON $ serialiseToTextEnvelope Nothing tx
+    txToJSON (TxInMode tx AllegraEraInCardanoMode) = toJSON $ serialiseToTextEnvelope Nothing tx
+    txToJSON (TxInMode tx MaryEraInCardanoMode) = toJSON $ serialiseToTextEnvelope Nothing tx
+    txToJSON (TxInMode tx AlonzoEraInCardanoMode) = toJSON $ serialiseToTextEnvelope Nothing tx
+    txToJSON (TxInMode tx BabbageEraInCardanoMode) = toJSON $ serialiseToTextEnvelope Nothing tx
+    txToJSON _ = String "<<a transaction>>"
+getNodeClientSelectorConfig Query =
+  SelectorConfig "query" True $ singletonFieldConfigWith (SomeJSON . queryToJSON) "query" True
+  where
+    queryToJSON :: Q.Some (QueryInMode CardanoMode) -> Value
+    queryToJSON (Q.Some query) = String . T.pack $ show query
+
+
+nodeClient :: Component IO (NodeClientDependencies r) NodeClient
+nodeClient =
+  component \NodeClientDependencies{..} ->
+    do
+      queryChannel <- newTChan
+      submitChannel <- newTChan
+      let
+        runNodeClient = mask \restore ->
+          do
+            ev <- newEvent eventBackend Connect
+            result <-
+              try @SomeException
+                . restore
+                $ connectToLocalNode LocalNodeClientProtocols
+                  {
+                    -- We don't use the block and tip information, but the chain
+                    -- sync client keeps the connection open. Without this client,
+                    -- the node would close the connection in about three seconds,
+                    -- making it impossible to submit and query.
+                    localChainSyncClient = LocalChainSyncClient chainSyncClient
+                  , localTxSubmissionClient = Just $ submitClient submitChannel
+                  , localTxMonitoringClient = Nothing
+                  , localStateQueryClient = Just $ queryClient queryChannel
+                  }
+            case result of
+              Left ex -> failEvent ev ex *> throw ex
+              Right _ -> pure ()
```

Comment:
This will keep the `Connect` event active for the entire run of the app, are you sure this is what we want? If it is, you can use `withEvent`
```suggestion
        runNodeClient = withEvent eventBackend Connect \_ -> connectToLocalNode LocalNodeClientProtocols
          {
            -- We don't use the block and tip information, but the chain
            -- sync client keeps the connection open. Without this client,
            -- the node would close the connection in about three seconds,
            -- making it impossible to submit and query.
            localChainSyncClient = LocalChainSyncClient chainSyncClient
          , localTxSubmissionClient = Just $ submitClient submitChannel
          , localTxMonitoringClient = Nothing
          , localStateQueryClient = Just $ queryClient queryChannel
          }
```

Reply:
I'll make this change. I actually just adopted the strategy that `marlowe-chain-indexer` was using: https://github.com/input-output-hk/marlowe-cardano/blob/1e778adb7f1b8e5087bdddabb034208626136ced/marlowe-chain-sync/chain-indexer/Language/Marlowe/Runtime/ChainIndexer/NodeClient.hs#L152.

Reply:
Fixed in c642758800a2cd902b0e2a324ffb8ded6211c738.